emc/dmri.py

```diff
@@ -0,0 +1,218 @@
+"""Representing data in hard-disk and memory."""
+from pathlib import Path
+from collections import namedtuple
+from tempfile import mkdtemp
+import attr
+import numpy as np
+import h5py
+import nibabel as nb
+from nitransforms.linear import Affine
+
+
+def _data_repr(value):
+    if value is None:
+        return "None"
+    return f"<{'x'.join(str(v) for v in value.shape)} ({value.dtype})>"
+
+
+@attr.s(slots=True)
+class DWI:
+    """Data representation structure for dMRI data."""
+
+    dataobj = attr.ib(default=None, repr=_data_repr)
+    """A numpy ndarray object for the data array, without *b=0* volumes."""
+    affine = attr.ib(default=None, repr=_data_repr)
+    """Best affine for RAS-to-voxel conversion of coordinates (NIfTI header)."""
+    brainmask = attr.ib(default=None, repr=_data_repr)
+    """A boolean ndarray object containing a corresponding brainmask."""
+    bzero = attr.ib(default=None, repr=_data_repr)
+    """
+    A *b=0* reference map, preferably obtained by some smart averaging.
+    If the :math:`B_0` fieldmap is set, this *b=0* reference map should also
+    be unwarped.
+    """
+    gradients = attr.ib(default=None, repr=_data_repr)
+    """A 2D numpy array of the gradient table in RAS+B format."""
+    em_affines = attr.ib(default=None)
+    """
+    List of :obj:`nitransforms.linear.Affine` objects that bring
+    DWIs (i.e., no b=0) into alignment.
+    """
+    fieldmap = attr.ib(default=None, repr=_data_repr)
+    """A 3D displacements field to unwarp susceptibility distortions."""
+    _filepath = attr.ib(default=Path(mkdtemp()) / "em_cache.h5", repr=False)
+    """A path to an HDF5 file to store the whole dataset."""
+
+    def __len__(self):
+        """Obtain the number of high-*b* orientations."""
+        return self.gradients.shape[-1]
+
+    def logo_split(self, index, with_b0=False):
+        """
+        Produce one fold of LOGO (leave-one-gradient-out).
+
+        Parameters
+        ----------
+        index : :obj:`int`
+            Index of the DWI orientation to be left out in this fold.
+
+        Return
+        ------
+        (train_data, train_gradients) : :obj:`tuple`
+            Training DWI and corresponding gradients.
+            Training data/gradients come **from the updated dataset**.
+        (test_data, test_gradients) :obj:`tuple`
+            Test 3D map (one DWI orientation) and corresponding b-vector/value.
+            The test data/gradient come **from the original dataset**.
+
+        """
+        if not Path(self._filepath).exists():
+            self.to_filename(self._filepath)
+
+        # read original DWI data & b-vector
+        with h5py.File(self._filepath, "r") as in_file:
+            root = in_file["/0"]
+            dwframe = np.asanyarray(root["dataobj"][..., index])
```

Comment:
curious -- why `dwframe` and `bframe` vs. just `test_data` and `test_gradients`?

Reply:
test_data and test_gradients imply there is some cross-validation going on. Because this is just a data representation object, it seemed more adequate not to refer to the framework you are working on.